dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp

```diff
@@ -631,6 +694,118 @@ std::unordered_map<TableID, SelectQueryInfo> DAGStorageInterpreter::generateSele
+        for (auto iter = regions_query_info.begin(); iter != regions_query_info.end(); /**/)
+        {
+            if (e.unavailable_region.find(iter->region_id) != e.unavailable_region.end())
+            {
+                // move the error regions info from `query_info.mvcc_query_info->regions_query_info` to `region_retry_from_local_region`
+                if (auto region_iter = dag_regions.find(iter->region_id); likely(region_iter != dag_regions.end()))
+                {
+                    region_retry_from_local_region.emplace_back(region_iter->second);
+                    buffer.fmtAppend("{},", region_iter->first);
+                }
+                iter = regions_query_info.erase(iter);
+            }
+            else
+            {
+                ++iter;
+            }
+        }
+        LOG_FMT_WARNING(
+            log,
+            "RegionException after read from storage, regions [{}], message: {}{}",
+            buffer.toString(),
+            e.message(),
+            (regions_query_info.empty() ? "" : ", retry to read from local"));
+        // no available region in local, break retry loop
+        // otherwise continue to retry read from local storage
+        return !regions_query_info.empty();
+    }
+    else
+    {
+        // push all regions to `region_retry_from_local_region` to retry from other tiflash nodes
+        for (const auto & region : query_info.mvcc_query_info->regions_query_info)
+        {
+            auto iter = dag_regions.find(region.region_id);
+            if (likely(iter != dag_regions.end()))
+            {
+                region_retry_from_local_region.emplace_back(iter->second);
+                buffer.fmtAppend("{},", iter->first);
+            }
+        }
+        LOG_FMT_WARNING(log, "RegionException after read from storage, regions [{}], message: {}", buffer.toString(), e.message());
+        return false; // break retry loop
+    }
+}
+
+void DAGStorageInterpreter::buildLocalStreamsForPhysicalTable(
+    const TableID & table_id,
+    const SelectQueryInfo & query_info,
+    DAGPipeline & pipeline,
+    size_t max_block_size)
+{
+    size_t region_num = query_info.mvcc_query_info->regions_query_info.size();
+    if (region_num == 0)
+        return;
+
+    assert(storages_with_structure_lock.find(table_id) != storages_with_structure_lock.end());
+    auto & storage = storages_with_structure_lock[table_id].storage;
+
+    const DAGContext & dag_context = *context.getDAGContext();
+    for (int num_allow_retry = 1; num_allow_retry >= 0; --num_allow_retry)
+    {
+        try
+        {
+            QueryProcessingStage::Enum from_stage = QueryProcessingStage::FetchColumns;
+            pipeline.streams = storage->read(required_columns, query_info, context, from_stage, max_block_size, max_streams);
+            injectFallPointForLocalRead(query_info);
+            validateQueryInfo(*query_info.mvcc_query_info, learner_read_snapshot, tmt, log);
```

Comment:
```suggestion
            // After getting streams from storage, we need to validate whether Regions have changed or not after learner read.
            // (by calling `validateQueryInfo`). In case the key ranges of Regions have changed (Region merge/split), those `streams`
            // may contain different data other than expected.
            validateQueryInfo(*query_info.mvcc_query_info, learner_read_snapshot, tmt, log);
```

Better to move these comments from `injectFailPointForLocalRead` to here. It explains why we need to call `validateQueryInfo` here

Reply:
my mistake..updtaed.